Groups and Permissions for Organization member

Rieven opened the pull request:
## Changes

- Add groups to the organization member model
- Migrate/update groups from the user model to the organization member model
- Clear all groups from the user model
- Permissions are set on groups, so check permissions using the organization member groups instead of user groups.
- Change PermissionrequiredMixin to our own custom RockyPermissionMixin to check permissions not only based on users but also for organization members.
- Change all views needing permissions to use the RockyPermissionMixin
- Change template based permissions to use our own custom template tag to check also for organization member permissions.
- Permissions set on user level overrule organization member permissions

## Issue ticket number and link
https://github.com/minvws/nl-kat-coordination/issues/582

## Proof
Groups in member
<img width="1109" alt="image" src="https://user-images.githubusercontent.com/30832687/228575956-06b8bf8d-ecd4-4cde-886e-fac95def3696.png">

Group in user, same as member:
<img width="1109" alt="image" src="https://user-images.githubusercontent.com/30832687/228576316-ef351bf2-fd26-46c7-8e16-5d86bf8f913c.png">


## Extra instructions for others
_This section may be skipped or omitted. Uncomment and answer the below questions if relevant._

<!---
- Does this PR introduce or depend on API-incompatible changes? If yes: what do other users/developers need to do or confirm before merging?
- Does this PR depend on a specific version of a library?
- Does this PR depend on any other pending PR's?
- Does this PR require config, setup, or `.env` changes?
-->

## Checklist for author(s):
- [x] All the commits in this PR are properly PGP-signed and verified;
- [x] This PR comes from a `feature` or `hotfix` branch, in line with our git branching strategy;
- [x] This PR is "bite-sized" and only focuses on a single issue, problem, or feature;
- [x] I am not reinventing the wheel: there is no high-quality library that already has this feature;
- [x] I have changed the example `.env` files if I added, removed, or changed any config options, and I have informed others that they need to modify their `.env` files if required;
- [x] I have performed a self-review of my own code;
- [x] I have commented my code, particularly in hard-to-understand areas;
- [x] I have made corresponding changes to the documentation, if necessary;
- [x] I have written unit, integration, and end-to-end tests for the change that I made;

If a non-trivial PR:
- [x] This PR is part of a milestone and has appropriate labels;
- [x] This PR is properly linked to the project board (either directly or via an issue);
- [x] I have added screenshots or some other proof that my code does what it is supposed to do;


```
## Checklist for functional reviewer(s):
- [ ] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [ ] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [ ] I confirmed that there are no unintended functional regressions in this branch;

### What works:
* _bullet point + screenshot (if useful) per tested functionality_

### What doesn't work:
* _bullet point + screenshot (if useful) per tested functionality_

### Bug or feature?:
* _bullet point + screenshot (if useful) if it is unclear whether something is a bug or an intended feature._
```

```
## Checklist for code reviewer(s):
- [ ] The code passes the CI tests and linters;
- [ ] The code does not bypass authentication or security mechanisms;
- [ ] The code does not introduce any dependency on a library that has not been properly vetted;
- [ ] The code does not violate Model-View-Template and our other architectural principles;
- [ ] The code contains docstrings, comments, and documentation where needed;
- [ ] The code prioritizes readability over performance where appropriate;
- [ ] The code conforms to our agreed coding standards.
```

TwistMeister commented on rocky/tools/models.py:
```diff
@@ -161,6 +161,7 @@ class STATUSES(models.TextChoices):
 
     user = models.ForeignKey("account.KATUser", on_delete=models.PROTECT, related_name="members")
     organization = models.ForeignKey(Organization, on_delete=models.CASCADE, related_name="members")
+    groups = models.ManyToManyField(Group, related_name="members", blank=True)
```
I feel like it would be logical that this defaults to 'client', since that's the group with the least permissions, and I think groups should never be empty. But I'm curious what @dekkers, @underdarknl or @praseodym think of this.

praseodym replied:
Why don't we remove `blank=True` and always require a group to be set? I don't think we'd need a default, we should be able to ensure a group is always set.

TwistMeister replied:
Yes, that's maybe even better. That enforces the user to make a decision. I like that suggestion.

Rieven replied:
Here is decided that we can have a none group member.

ammar92 commented on rocky/tools/models.py:
```diff
@@ -170,6 +182,55 @@ class STATUSES(models.TextChoices):
+            perms = permissions
+        for perm in perms:
+            try:
+                _, codename = perm.split(".", 1)
```
Wouldn't counting the number of dots be simpler?
```if perm.count(".") != 1: raise ValueError```

ammar92 commented on rocky/tools/models.py:
```diff
@@ -170,6 +182,55 @@ class STATUSES(models.TextChoices):
+    def is_client(self):
+        return self.groups.filter(name=GROUP_CLIENT).exists()
+
+    def get_permissions_codenames(self, permissions):
```
Where are the tests for this? Based on just the code I can see that `"app_label."` also works, which contradicts with the error message

ammar92 commented on rocky/tools/models.py:
```diff
@@ -170,6 +182,55 @@ class STATUSES(models.TextChoices):
     def blocked(self):
         return self.status == OrganizationMember.STATUSES.BLOCKED
 
+    @property
+    def is_admin(self):
+        return self.groups.filter(name=GROUP_ADMIN).exists()
+
+    @property
+    def is_redteam(self):
+        return self.groups.filter(name=GROUP_REDTEAM).exists()
+
+    @property
+    def is_client(self):
+        return self.groups.filter(name=GROUP_CLIENT).exists()
```
I think these properties should be cached. Also, are they still necessary since with this approach we don't have to check for fixed groups anymore but permissions?

Rieven replied:
Indeed caching here is a good idea. We use these props in just a few cases where we want to know if a user is in a particular group for redirecting purposes, mostly used in onboarding.

dekkers replied:
Those checks should be changed, because we should not hardcode these groups and that means those groups don't have to exist.

Rieven replied:
Here decided that we are going to fix this in another PR. We have to think about the design of how we are going to use these groups in the future.

Darwinkel commented:
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make kat`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

### What works:
* No apparent regressions
* No issues with existing user management

dekkers commented on rocky/tools/models.py:
```diff
@@ -157,6 +171,7 @@ class STATUSES(models.TextChoices):
 
     user = models.ForeignKey("account.KATUser", on_delete=models.PROTECT, related_name="members")
     organization = models.ForeignKey(Organization, on_delete=models.CASCADE, related_name="members")
+    groups = models.ManyToManyField(Group, related_name="members", blank=True)
```
I think "members" as related name for groups is confusing as it won't really give the members of the group, but the relation between organization, user and group. I think the default organizationmember_set would be fine if we would use it in the code.

dekkers commented on rocky/tools/models.py:
```diff
@@ -170,6 +185,55 @@ class STATUSES(models.TextChoices):
+
+        for codename in codenames:
+            for group in self.groups.all():
+                group_permissions_codenames = group.permissions.values_list("codename", flat=True)
```
This is wrong. Both the app and codename should match. The permission "app1.can_do_something" is not the same as a permission "app2.can_do_something". It also does too many queries.

We can get all the permissions with a single query:

```
perms = ["%s.%s" % (ct, name) for ct, name in Permission.objects.filter(group__organizationmember=member).values_list("content_type__app_label", "codename")]
```

dekkers commented on rocky/tools/migrations/0034_update_perms_move_and_clear_groups.py:
```diff
@@ -0,0 +1,55 @@
+    for app_config in apps.get_app_configs():
+        app_config.models_module = True
+        create_permissions(app_config, apps=apps, verbosity=0)
+        app_config.models_module = None
```
There is no need for this in this migration as far as I can see. Either the admin and redteam group exists, but then we have already run migrate before and the permissions have been created, or the groups don't exist but then we also don't need the permissions.

dekkers commented on rocky/tools/migrations/0034_update_perms_move_and_clear_groups.py:
```diff
@@ -0,0 +1,55 @@
+    for member in members:
+        user_groups = member.user.groups.all()
+        for group in user_groups:
+            member.groups.add(group)
```
This can be done in one query using:

`member.groups.add(*user_groups)`

Rieven commented on rocky/tools/models.py:
```diff
@@ -170,6 +185,55 @@ class STATUSES(models.TextChoices):
+    def has_member_perms(self, permissions: Union[str, tuple] = None) -> bool:
+        if self.user.is_superuser and self.user.is_active and self.status is not OrganizationMember.STATUSES.BLOCKED:
+            return True
+        codenames = self.get_permissions_codenames(permissions)
```
This is how it is implemented now. The permission parameter comes in as string or a tuple of string perms, the `get_permissions_codenames` will split the app_label and perm and see if it exists. According to your other comments this part will be refactored to a more simpler approach.

Rieven commented on rocky/account/templates/account_detail.html:
```diff
@@ -63,7 +64,8 @@ <h1>{% translate "Account details" %}</h1>
         </section>
-        {% if perms.tools.can_scan_organization %}
+        {% has_organization_perms "tools.can_scan_organization" request.user organization as can_scan_organization %}
+        {% if can_scan_organization %}
```
I have made a MemberPermissionMixin where we can pass it to the context as an instance like:  `perms = MemberPermissionMixin()`

then we can do `perms.has_member_perms("tools.can_scan_organization", member)`

dekkers commented:
I implemented organization member permissions check in the templates in the same way as django itself does using the `perms` context variable. In the OrganizationView we set the `perms` context variable which will overrule the context processor. This means that we can just use `perms` everywhere and for views that inherit from OrganizationView this means that the permissions from the OrganizationMember groups are added to the perms.

I also moved the organization member permissions checks to the OrganizationMember model as this is where they belong. This makes it also easy to cache the permissions similar as how it done on the User model to minimize database queries.

I also renamed `RockyPermissionRequiredMixin` to `OrganizationPermissionRequiredMixin` as the mixin is specific to the organization views and not for all of rocky.

Rieven commented:
> I implemented organization member permissions check in the templates in the same way as django itself does using the `perms` context variable. In the OrganizationView we set the `perms` context variable which will overrule the context processor. This means that we can just use `perms` everywhere and for views that inherit from OrganizationView this means that the permissions from the OrganizationMember groups are added to the perms.
> 
> I also moved the organization member permissions checks to the OrganizationMember model as this is where they belong. This makes it also easy to cache the permissions similar as how it done on the User model to minimize database queries.
> 
> I also renamed `RockyPermissionRequiredMixin` to `OrganizationPermissionRequiredMixin` as the mixin is specific to the organization views and not for all of rocky.

Very well implemented!

praseodym commented:
In the members view I can no longer see which role a member has:
![image](https://user-images.githubusercontent.com/20283/235471904-32ba5d84-c5f1-4162-8d39-85448ee169b5.png)

Rieven commented:
> In the members view I can no longer see which role a member has: ![image](https://user-images.githubusercontent.com/20283/235471904-32ba5d84-c5f1-4162-8d39-85448ee169b5.png)

fixed in commit: https://github.com/minvws/nl-kat-coordination/pull/606/commits/fb293e068a6887fbdbb5ab4c4f668a9240701877

praseodym commented:
lgtm!

Darwinkel commented:
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make kat`.
- [ ] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.


### What doesn't work:
* After creating a new redteamer, logging in, and doing the onboarding:
![image](https://user-images.githubusercontent.com/9536436/235614838-78e9bb78-128e-40b6-8ce0-e54aea8575fe.png)
* As a redteamer without any clearance, I can start L1+ scans. This issue remains after an admin has modified the assigned clearance level. In other words: authorization is broken. 
* To make matters worse, I can, as a client, initiate scans.

### Bug or feature?:
* Not sure if redteamers should be onboarded if being created by an admin. Currently, logging in as a redteamer triggers a flow which encourages a new user to add a new object.

Rieven commented:
As a redteamer without any clearance, I can start L1+ scans. This issue remains after an admin has modified the assigned clearance level. In other words: authorization is broken.

@Darwinkel 
All and all good catch! This problem is not directly related to this feature as it is also broken in main. The problem here is that if admin creates a new redteam member without setting clearance levels, then a redteam will encounter this problem anyways, because there is no acknowledgement for a scan level L2.  No levels are set for this new member. For the onboarding purpose we need to create an extra screen where the redteamer must acknowledge the L2 to scan for DNS. 

This is something not yet implemented and should have been thought off in the first place.

Darwinkel commented:
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make kat`.
- [ ] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.


### What works:
- Clients can no longer initiate scans. I talked with @TwistMeister and there is something to be said for a client being able to run (some) scans, as this is also what the scheduler does in the background. It may be relevant for them to generate a more recent report with the newest information. The same essentially goes for low-clearance redteamers: shouldn't they be able to perform manual scans as long as this is something that the scheduler would trigger by itself?

### What doesn't work:
* https://github.com/minvws/nl-kat-coordination/issues/915
* https://github.com/minvws/nl-kat-coordination/pull/783
* https://github.com/minvws/nl-kat-coordination/issues/908

Rieven commented:
Shouldn't they be able to perform manual scans as long as this is something that the scheduler would trigger by itself?

No, they shouldn't indeed....Clearance checks should be checked for manual scans.
